### ISSUE TITLE
ColumnFamilyHandle must be free'd before RocksDB in Java API

### ISSUE THREAD

**Issue by adamretter** [facebook/rocksdb#974]
Currently with the Java API, you have to explicitly free any `ColumnFamilyHandle` before you free the `RocksDB` object. This provides for an asymmetrical open/close pattern, i.e.:

``` java
final List<ColumnFamilyHandle> cfHandles = new ArrayList<>();
try (final RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath(), cfDescriptors, cfHandles)) {
  try {
    /* do stuff... */
  } finally {
    for (final ColumnFamilyHandle columnFamilyHandle : cfHandles) {
      columnFamilyHandle.close();
    }
  }
}
```

A more desirable symmetrical pattern would involved freeing any `ColumnFamilyHandle` after the `RocksDB` object. This seems to be possible using the C++ example given in https://github.com/facebook/rocksdb/blob/master/examples/column_families_example.cc#L31. The Java approach would then look like:

``` java
final List<ColumnFamilyHandle> cfHandles = new ArrayList<>();
try (final RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath(), cfDescriptors, cfHandles)) {
    /* do stuff... */
} finally {
  for (final ColumnFamilyHandle columnFamilyHandle : cfHandles) {
    columnFamilyHandle.close();
  }
}
```

Perhaps an even nicer approach for the Java API would be to associate any `ColumnFamilyHandle` with the `RocksDB` object, and then closing the `RocksDB` object would also take care of freeing any `ColumnFamilyHandle` object too. Which would leave us with the easiest approach for Java users, e.g.:

``` java
final List<ColumnFamilyHandle> cfHandles = new ArrayList<>();
try (final RocksDB db = RocksDB.open(options, dbFolder.getRoot().getAbsolutePath(), cfDescriptors, cfHandles)) {
    /* do stuff... */
}
```

@yhchiang Any comments?

**Comment by yhchiang** [comment]
Yep, I think associating ColumnFamilyHandle's life cycle with RocksDB makes sense, and it would be easier to use.

**Comment by gfosco** [comment]
Closing this via automation due to lack of activity.  If discussion is still needed here, please re-open or create a new/updated issue.

**Comment by xin-au** [comment]
Wondering if I use `public static RocksDB open(final String path)`, there are no arguments I could close them, should I just call `rocksDB.close()`?

**Comment by adamretter** [comment]
@xin-au Not sure I understand. If you open RocksDB like that, there are no column families, so you only need to call rocksdb.close

**Comment by xin-au** [comment]
@adamretter , thanks for your response! And sorry for my unclear words. I just used function `public static RocksDB open(final String path)` without `column families`. And when I run some app tests, I noticed that seems my rocksdb not close properly, I did close the db and close the iterator, but I got the error :
```
Caused by: org.rocksdb.RocksDBException: While lock file: name-store/LOCK: No locks available
```
What do you think? Is there anything I need to be aware when I close rocksdb?

**Comment by adamretter** [comment]
@alanpaxton I can confirm that this is addressed in #7428
